File: 1615-maximal-network-rank/1615-maximal-network-rank.cpp
class Solution {
public:
    int maximalNetworkRank(int n, vector<vector<int>>& roads) {
        vector<set<int>> graph(n);
        for (auto r : roads) {
            graph[r[0]].insert(r[1]);
            graph[r[1]].insert(r[0]);
        }
        int res = 0;
        for (int a = 0; a < n; a++) {
            for (int b = a + 1; b < n; b++) {
                int rank = graph[a].size() + graph[b].size();
                if (graph[a].count(b)) {
                    rank--;
                }
                res = max(res, rank);
            }
            
        }
        return res;
    }
};

File: 1615-maximal-network-rank/1615-maximal-network-rank.py
class Solution:
    def maximalNetworkRank(self, n: int, roads: List[List[int]]) -> int:
        adj = defaultdict(lambda: set())
        for u, v in roads:
            adj[u].add(v)
            adj[v].add(u)
        res = 0
        for u in range(n):
            for v in range(u + 1, n):
                rank = len(adj[u]) + len(adj[v])
                if v in adj[u]:
                    rank -= 1
                res = max(res, rank)
        return res